File: oj/projecteuler/343.cc
#include <bits/stdc++.h>

using namespace std;

using ull = unsigned long long;

vector<ull> primes;

ull get_max_factor(ull x) {
  ull ele = 1;
  for(auto i : primes) {
    if(i > x) break;
    if((x % i) == 0) {
      ele = i;
      while(0 == (x%i)){
	x /= i;
      }
    }
  }
  return ele;
}

ull f(ull i) {
  return max(get_max_factor(i+1), get_max_factor(i*i - i + 1)) - 1;
}

int main(){
  vector<ull> all(2000000, 0);
  for(int i = 2; i < 2000000; i++) {
    if(all[i] == 0) {
      primes.push_back(i);
      int j = i;
      while(j < 2000000) {
	all[j] = 1;
	j += i;
      }
    }
  }
  
  /*
  unsigned long long i = 2000000;
  cout<<get_max_factor(i*i*i+1)<<endl;
  unsigned long long  sum = 0;
  for(unsigned i = 1; i <= 100; i++) {
    sum += f(i);
  }
  

  cout<<sum<<endl;
  */
  ull sum = 0; 
  for(unsigned i = 2000000; i >= 1; i--) {
    cout<< f(i)<<endl;

    if(i % 10000 == 0) {
      cerr<<"running... "<<i<<endl;
    }
  }

  cout<<sum<<endl;
  
}

File: oj/projecteuler/343.go
package main

import (
       "fmt"
       "sync"
)

func main() {
  var primes []uint64
  var all [2000000]uint64

  all[0] = 1
  all[1] = 1
  for i, _ := range all {
    if all[i] == 0 {
      primes = append(primes, uint64(i))
      j := i
      for j < 2000000 {
        all[j] = 1
        j += i
      }
    }
  }

  for _, i := range primes {
    fmt.Println(i)
    break
  }

  //var sum uint64 = 0
  var wg sync.WaitGroup
  var i uint64 = 1
  messages := make(chan uint64)
  for i = 1; i <= 2000000; i++ {
    wg.Add(1)
    go func(k uint64){
      defer wg.Done()
      var a uint64 = k + 1

      var ele uint64 = 1
      for _, p := range primes {
        if p > a {
          break
        }

        if (a % p) == 0 {
          ele = p
          for 0 == (a % p) {
            a /= p
          }
        }
      }
      ea := ele

      a = k*k - k + 1
      ele = 1
      for _, p := range primes {
        if p > a {
          break
        }

        if (a % p) == 0 {
          ele = p
          for 0 == (a % p) {
            a /= p
          }
        }
      }
      eb := ele
      if ea > eb {
        messages <- ea
      } else {
        messages <- eb
      }
    }(i)
  }

  go func() {
    var x uint64
    var sum uint64
    for i := range messages {
      //fmt.Println(i)
      if (x % 10000)  == 0{
        fmt.Println(x)
      }
      x += 1
      sum += i
    }
    fmt.Println(sum)
  }()
  wg.Wait()
}
